sdk/clientcore/core/src/main/java/io/clientcore/core/util/Union.java

```diff
@@ -0,0 +1,285 @@
+            }
+        }
+        throw LOGGER.logThrowableAsError(new IllegalArgumentException("Invalid type: " + value.getClass().getName()));
+    }
+
+    /**
+     * Gets the type of the value.
+     *
+     * @return The type of the value.
+     */
+    public Type getCurrentType() {
+        return currentType;
+    }
+
+    /**
+     * Gets the types of the union.
+     *
+     * @return The types of the union.
+     */
+    public List<Type> getSupportedTypes() {
+        return Collections.unmodifiableList(types);
+    }
+
+    /**
+     * Gets the value of the union.
+     *
+     * @return The value of the union.
+     * @param <T> The type of the value.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T getValue() {
+        return (T) value;
+    }
+
+    /**
+     * Gets the value of the union if it is of the expected type.
+     *
+     * @param clazz The expected type of the value.
+     * @return The value of the union.
+     * @param <T> The expected type of the value.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T getValue(Class<T> clazz) {
+        if (clazz.isInstance(value)) {
+            return clazz.cast(value);
+        }
+        if (isPrimitiveTypeMatch(value, clazz)) {
+            return (T) value;
+        }
+        throw LOGGER.logThrowableAsError(new IllegalArgumentException("Value is not of type: " + clazz.getName()));
+    }
+
+    /**
+     * Gets the value of the union if it is of the expected type.
+     *
+     * @param clazz The expected type of the value.
+     * @param genericTypes The generic types of the expected type.
+     *
+     * @return The value of the union.
+     * @param <T> The expected type of the value.
+     */
+    public <T> T getValue(Class<T> clazz, Class<?>... genericTypes) {
+        return getValue(new ParameterizedTypeImpl(clazz, genericTypes));
+    }
+
+    /**
+     * Gets the value of the union if it is of the expected type.
+     *
+     * @param type The expected type of the value.
+     *
+     * @return The value of the union.
+     * @param <T> The expected type of the value.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T getValue(ParameterizedType type) {
+        if (isInstanceOfType(value, type)) {
+            return (T) value;
+        }
+        throw LOGGER.logThrowableAsError(new IllegalArgumentException("Value is not of type: " + type.getTypeName()));
+    }
+
+    /**
+     * This method is used to consume the value of the Union if it is of the expected type.
+     *
+     * @param consumer A consumer that will consume the value of the Union if it is of the expected type.
+     * @param clazz The expected type of the value.
+     * @return Returns true if the value was consumable by the consumer, and false if it was not.
+     * @param <T> The value type expected by the consumer.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> boolean tryConsume(Consumer<T> consumer, Class<T> clazz) {
```

Comment:
In all the cases where we are getting or consuming we should have a quick path check of `clazz` / `type` == `currentType`.

Reply:
Should we support chaining of consumers like we have in Java Functions, Predicate etc?

```java
union.tryConsume(stringConsumer, String.class)
         .ifNot(IntegerConsumer, Integer.class)
         .ifNot(BooleanConsumer, Boolean.class); 
```

Reply:
It will be like 

```
    public <T> Union tryConsume(Consumer<T> consumer, Class<T> clazz) {
        if (clazz == currentType) {
            consumer.accept((T) value);
        } else if (isInstanceOfType(value, clazz)) {
            consumer.accept(clazz.cast(value));
        } else if (isPrimitiveTypeMatch(value, clazz)) {
            consumer.accept((T) value);
        }
        return this;
    }

    public <T> Union ifNot(Consumer<T> consumer, Class<T> clazz) {
        tryConsume(consumer, clazz);
        return this;
    }
```

and sample 

```
  union.setValue(1)
            .tryConsume(v -> System.out.println("String value: " + v), String.class)
            .ifNot(v -> System.out.println("Integer value: " + v), Integer.class);
```

Reply:
https://github.com/Azure/azure-sdk-for-java/issues/43953